### PR TITLE
Modinvosto 2 Define APIs

### PR DESCRIPTION
**Pull request by pautau-s**
## Purpose
[MODINVOSTO-2](https://issues.folio.org/browse/MODINVOSTO-2) Create schemas and samples for invoice and invoiceLine

## Approach
- created schemas for invoice, collection of invoices, invoice line, collection of invoice lines based on [schema details document](https://docs.google.com/spreadsheets/d/1F88PoSnEKD7oFnza1ZuOgzAgKnK95XOyTQzxyww3tZA/edit?ts=5c1bd54f#gid=131223376)
- created samples for these schemas
- moved schema and sample for sequence number to common folder
- moved schemas and sample for composite purchase order and composite poLine to mod-orders folder

#### TODOS and Open Questions
- [ ] Update API test for mod-orders

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Examples exist for all schemas
  - [x] Descriptions exist for all schema properties
  - [x] All schemas pass raml-cop linting
- Does this introduce breaking changes?
  - [ ] Were there any schema changes?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: mod-orders/schemas/composite_po_line.json [hunk 1]
```diff
@@ -44,7 +44,7 @@
     "adjustment": {
       "description": "the adjustment record associated with this purchase order line",
       "type": "object",
-      "$ref": "mod-orders-storage/schemas/adjustment.json"
+      "$ref": "../../mod-orders-storage/schemas/adjustment.json"
```
Comment (craigmcnally) [inline mod-orders/schemas/composite_po_line.json]: Have you tried this out?  There was at one point a technical reason why the composite schemas needed to live at the top level... something like RMB couldn't resolve the references...

Comment (pautau-s) [reply mod-orders/schemas/composite_po_line.json]: It works. I checked three times: for mod-invoice-storage, mod-orders-storage, mod-orders.
I think it was fixed with [RMB-203](https://issues.folio.org/browse/RMB-203)

Comment (craigmcnally) [reply mod-orders/schemas/composite_po_line.json]: Sweet, good to know!

File: mod-invoice-storage/examples/invoice_line.sample [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  "comment": "Sample invoice line",
+  "price": 2.2,
+  "quantity": 3.0,
+  "doNotReleaseEncumbrance": true,
+  "reportTax": 3.5,
+  "subscriptionInfo": "Subscription information",
+  "subscriptionStart": "2018-08-01T00:00:00.000+0000",
+  "subscriptionEnd": "2019-01-01T00:00:00.000+0000",
+  "pieceId": "5e317dc2-deeb-4429-b2a1-91e5cd0fd5f7",
+  "invoiceUuid": "c0d08448-347b-418a-8c2f-5fb50248d67e",
```
Comment (craigmcnally) [inline mod-invoice-storage/examples/invoice_line.sample]: For consistency, let's change this to `invoiceId`

File: mod-invoice-storage/examples/invoice_line.sample [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  "invoiceLineTax": 11.2,
+  "comment": "Sample invoice line",
+  "price": 2.2,
+  "quantity": 3.0,
+  "doNotReleaseEncumbrance": true,
+  "reportTax": 3.5,
+  "subscriptionInfo": "Subscription information",
+  "subscriptionStart": "2018-08-01T00:00:00.000+0000",
+  "subscriptionEnd": "2019-01-01T00:00:00.000+0000",
+  "pieceId": "5e317dc2-deeb-4429-b2a1-91e5cd0fd5f7",
```
Comment (craigmcnally) [inline mod-invoice-storage/examples/invoice_line.sample]: I think this should be an array (is repeatable).   Also, lets make it plural `pieceIds`  I'll ask Dennis to update the spreadsheet

File: mod-invoice-storage/examples/invoice_line.sample [hunk 4]
```diff
@@ -0,0 +1,37 @@
+{
+  "id": "e0d08448-343b-118a-8c2f-4fb50248d672",
+  "description": "Some description",
+  "poLineId": [
```
Comment (craigmcnally) [inline mod-invoice-storage/examples/invoice_line.sample]: `poLineIds` (plural) - Updated today by Dennis.

File: mod-invoice-storage/examples/invoice_collection.sample [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    "prorata": false,
+    "status": "Open",
+    "source": "024b6f41-c5c6-4280-858e-33fba452a334",
+    "totalAmount": 5,
+    "vendorInvoiceNo": "YK75851",
+    "disbursementNumber": "58",
+    "voucherNumber": "478",
+    "paymentId": "a5065f0d-fb88-4d23-b0c1-57e754fba40e",
+    "disbursementDate": "2018-08-19T00:00:00.000+0000",
+    "poId": [
```
Comment (craigmcnally) [inline mod-invoice-storage/examples/invoice_collection.sample]: This has changed to `poNumbers` string (not UUID, but an array of purchase order numbers).  Was just updated in the spreadsheet today.

File: mod-invoice-storage/examples/invoice_line.sample (hunk 2 above)
Comment (pautau-s) [reply mod-invoice-storage/examples/invoice_line.sample]: Done

File: mod-invoice-storage/examples/invoice_line.sample (hunk 3 above)
Comment (pautau-s) [reply mod-invoice-storage/examples/invoice_line.sample]: Updated

File: mod-invoice-storage/examples/invoice_line.sample (hunk 4 above)
Comment (pautau-s) [reply mod-invoice-storage/examples/invoice_line.sample]: Updated

File: mod-invoice-storage/examples/invoice_collection.sample (hunk 5 above)
Comment (pautau-s) [reply mod-invoice-storage/examples/invoice_collection.sample]: Changed

File: mod-invoice-storage/schemas/invoice.json [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+      "description": "This does not denote a user ID but describes how the record was created. Eg. Manually, EDIFACT, MARC etc",
+      "type": "string",
+      "pattern": "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$"
+    },
+    "totalAmount": {
+      "description": "Total amount",
+      "type": "integer"
+    },
+    "vendorInvoiceNo": {
+      "description": "This is the number from the vendor's invoice this may or may not be the same as the invoice_no",
```
Comment (piotr-kalashuk) [inline mod-invoice-storage/schemas/invoice.json]: I believe invoice_no is actually folioInvoiceNo property

File: mod-invoice-storage/schemas/invoice.json [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      ]
+    },
+    "source": {
+      "description": "This does not denote a user ID but describes how the record was created. Eg. Manually, EDIFACT, MARC etc",
+      "type": "string",
+      "pattern": "^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$"
+    },
+    "totalAmount": {
+      "description": "Total amount",
+      "type": "integer"
```
Comment (piotr-kalashuk) [inline mod-invoice-storage/schemas/invoice.json]: I believe this should be defined as `number` like in the [invoice.schema](https://github.com/folio-org/deprecated-mod-acquisitions/blob/master/ramls/_schemas/invoice.schema)

File: mod-invoice-storage/schemas/invoice_line.json [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+      "type": "string"
+    },
+    "productId": {
+      "description": "product identifier",
+      "type": "string"
+    },
+    "productIdType": {
+      "description": "the type of id",
+      "type": "string",
+      "enum": [
```
Comment (piotr-kalashuk) [inline mod-invoice-storage/schemas/invoice_line.json]: Probably better to move to separate file in common place and refer here and in `details.json` of the PO Line schema.

File: mod-invoice-storage/schemas/invoice_line.json [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+      "description": "Free form commentary",
+      "type": "string"
+    },
+    "price": {
+      "description": "Invoice line price",
+      "type": "number"
+    },
+    "quantity": {
+      "description": "quantity",
+      "type": "number"
```
Comment (piotr-kalashuk) [inline mod-invoice-storage/schemas/invoice_line.json]: Probably better to define as `integer`

File: mod-invoice-storage/schemas/invoice_line.json [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    "doNotReleaseEncumbrance": {
+      "description": "If this is intended to not only release encumbrances but to indicate that this is the final invoice. Should the label be changed? This should take into account the fiscal year. We may release encumbrance for this FY but for on-going orders expect to encumber again next year.",
+      "type": "boolean"
+    },
+    "reportTax": {
+      "description": "Report tax",
+      "type": "number"
+    },
+    "subscriptionInfo": {
+      "description": "Information on the subscription",
```
Comment (piotr-kalashuk) [inline mod-invoice-storage/schemas/invoice_line.json]: Use from the doc
`the user should use this field to provide info on the subscription (eg. which volumes are being invoiced). It is MANDATORY if the chk_subscription_overlap is FALSE AND there is no start date provided`

File: mod-invoice-storage/schemas/invoice.json (hunk 6 above)
Comment (craigmcnally) [reply mod-invoice-storage/schemas/invoice.json]: Good catch

**Pull request merged**